WPMedia/fusion-news-theme-blocks
# PEN-1127: Main menu (hamburger) within tab order of page

**If you have not filled out the checklist below, the pr is not ready for review.**

## Description
Added FocusTrap NPM module to Header Nav Chain Block and implemented for section navigation menu. Also added logic to prevent section menu links from being focused when menu is closed.

## Jira Ticket
- [PEN-1127](https://arcpublishing.atlassian.net/browse/PEN-1127)

## Acceptance Criteria
![image](https://user-images.githubusercontent.com/26662906/108243788-17f2ba00-7114-11eb-9f8a-ac924397aa80.png)

## Test Steps
1. In `Fusion-News-Theme` repo, checkout `master` & `git pull`
2. After checking out this feature branch in `fusion-news-theme-blocks`, run `rm -rf ./node_modules && npx lerna clean && npm i && npx lerna bootstrap && cd blocks/header-nav-chain-block && npm i && cd ../..`
3. Run `npx fusion start -f -l @wpmedia/header-nav-chain-block,@wpmedia/shared-styles` in `Fusion-News-Theme`
4. On any page, setup Header Nav Chain Block to have "Section Menu" and "Arc Search" button on the left side of the nav bar. Stage, publish, open in separate tab.
5. On page load, tab through all nav bar elements until you reach the last one. Upon reaching this last link in the nav bar, you should be taken to the next focusable element on the visible page content. The focus should no longer move into the hidden section navigation menu.
6. Reload page, tab to the "Sections" button, click "ENTER" to open the section navigation menu. Once open, tab through the focusable elements in the section nav menu until you reach the end. Upon reaching the last focusable element, click the tab key again and you should be taken back to the first focusable element in the nav menu (see screenshot below).
7. Close section navigation menu panel using the "ESC" key and you should be taken back to the next focusable element in the nav bar.  

## Effect Of Changes
### Before
Links in section menu navigation were focusable even when the drawer wasn't open. When opening the navigation menu and tabbing through link elements, there was no isolation from the rest of the page so you would be able to "tab out" of the navigation menu once you reach the end of focusable elements.

### After
Now, a "focus trap" has been added around the section menu navigation container element which prevents focus from moving outside of the drawer until it is closed. When you reach the end of the list of focusable elements in the section menu navigation, it will simply start back at the first focusable element in the container. It no longer moves back to the main page content when reaching the end of the section menu links (tabbing through).

This code changes also creates isolation when the section menu navigation pane is closed. You will no longer be able to accidentally tab over to a hidden section menu link when the nav menu is not open.

![image](https://user-images.githubusercontent.com/26662906/108247430-4377a380-7118-11eb-892f-36dce2b23bae.png)

## Dependencies or Side Effects
_Examples of dependencies or side effects are:_
- Additional settings that will be required in the blocks.json
- Changes to the custom fields which will require users to reconfigure features
- Update to css framework or SDK
- Dependency on another PR that needs to be merged first

## Author Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
please explain why (so that we can fix it whenever it gets refactored).

## Reviewer Checklist 
_The reviewer of the PR should copy-paste this template into the review comments on review._
